### ISSUE TITLE
[Feature request]: support mpi backend in DistributedDataParallel

### ISSUE THREAD

**Issue by xhzhao** [pytorch/pytorch#5868]
AFAIK, PyTorch [DDP](http://pytorch.org/docs/master/nn.html#torch.nn.parallel.DistributedDataParallel) only support nccl and gloo backend, and i think it would be great to support mpi backend when the user get CPU only, especially for the researcher with supper computer access.

I tried to build a proto in this [github](https://github.com/xhzhao/PyTorch-MPI-DDP-example), and my example passed the accuracy validation on MNIST+LeNet and AN4+DeepSpeech2. Besides, this [distributed.py](https://github.com/xhzhao/PyTorch-MPI-DDP-example/blob/master/data/distributed.py) has exactly the same usage as the PyTorch DDP, so i'm here to check if i should propose a PR to merge the code to the PyTorch [distributed.py](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py). If yes, the PyTorch DDP will support both CPU and GPU with exactly the same interface. There maybe a bug in model saving, and i will get it fixed in the following days.

BTW, we found the async SGD is also very popular in multi-node training, and we have a plan to enable async SGD( say [elastic SGD](https://arxiv.org/abs/1412.6651)) on CPU side .

**Comment by ailzhang** [comment]
Hi @xhzhao, thanks for working on it! Yea we would love to have MPI backend support DDP, please go ahead and create a PR for it.
